### PR TITLE
Loopback and Pattern Generation

### PR DESCRIPTION
**Pull request by lsgunth**
This draft PR adds Loopback and Pattern Generation (SOW 3.1.5).

The loopback functionality seems to work, but the pattern generation still gives odd behaviour and may need some changes. 

The PR adds the "loopback" and "pattern" diag commands. See their respective helps for details.

### CODE REVIEW

**Comment by yungwesl** [comment]
Hi Logan,

I have verified the loopback code is working as expected.  There are some unexpected behaviors on the pattern generator which I will continue to look into. 

From reviewing the MRPC implementation there is a disconnected between the way the MRPC input/output is organized and how the underlying hardware works.
- Pattern generation and monitoring enables are port-based which is aligned between HW and MRPC
- Error checking is lane-based (not port based) which is not aligned between HW and MRPC
- Error injection is lane-based (not port based) which is not aligned between HW and MRPC

As a result with the current limitations of the MRPC command interface, switchtec_user will be limited in the following ways
- Pattern generation and monitor enables will continue to be port based
- Error checking only reports the first lane's data (if you have some time can you try reading MRPC offset 0x80C and beyond? 0x804 and 0x808 should provide lower/upper counts of lane 0, logically 0x80C and 0x810 should report lane 1 etc)
- Error injection is limited to injecting a fixed number of errors across all lanes in a given port

Wes

**Comment by lsgunth** [comment]
I can try reading those next week.

But as I mentioned in email, I started with a design that allowed port_ids that were unaligned, but the hardware always returned an error when I tried to send them. So I switched to a design that was port based.

**Comment by yungwesl** [comment]
Some additional implementation details:

The sequence for a loopback test using Rx to Tx loopback
(0) A connection is made between Port A and Port B
(1) Make sure the bifurcation of the port matches the link width you are testing (using a x4 connection on a x16 bifurcation may result in unrelated errors)
(2) Establish LTSSM L0 on Port A and B
(3) Enable Loopback Rx to Tx on Port A
(4) Enable pattern generation on Port B
(5) Enable pattern monitoring on Port B
(6) Check for errors on Port B

The pattern checking in the switchtec_user implementation does appear to be different than the UART CLI.
e.g.
(1) use switchtec_user diag loopback to enable Rx-Tx loopback on Port A
(2) use switchtec_user diag pattern to generate PRBS7 on port B
(3) use switchtec_user diag pattern to monitor PRBS7 on Port B
(4) in UART CLI run the command 
{code}
patgenchk -chk -get <port B> <lane 0>
{code}
Check if the numbers match
(5) run patgenchk -err -inj <port A> 3 to inject 3 errors on all lanes in the port
(6) run patgenchk -chk -get <port B> <lane 0> to confirm 3 errros
(7) check this against switchtec user pattern -m

**Comment by yungwesl** [comment]
Here is an example use case using purely UART CLI commands (for which the MRPC are supposed to be identical).
---
0x00000000:0000>lnkstat
[00] part:00.01 w:cfg[x08]-neg[x04] stk:0.0 port_active_lane:0123xxxx         dsp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[04] part:01.01 w:cfg[x08]-neg[x00] stk:0.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[08] part:00.02 w:cfg[x16]-neg[x00] stk:1.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[16] part:00.03 w:cfg[x08]-neg[x04] stk:2.0 port_active_lane:0123xxxx         dsp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[20] part:01.02 w:cfg[x08]-neg[x00] stk:2.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[24] part:00.04 w:cfg[x08]-neg[x04] stk:3.0 port_active_lane:0123xxxx         dsp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[28] part:01.03 w:cfg[x08]-neg[x00] stk:3.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[32] part:01.00 w:cfg[x16]-neg[x04] stk:4.0 port_active_lane:0123xxxxxxxxxxxx usp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[40] part:00.00 w:cfg[x16]-neg[x08] stk:5.0 port_active_lane:01234567xxxxxxxx usp dl_active:1 Rate:G3 LTSSM: L0     -L0          
[48] part:00.05 w:cfg[x04]-neg[x04] stk:6.0 port_active_lane:0123             dsp dl_active:1 Rate:G4 LTSSM: L0     -L0

// SETS THE LOOPBACK ON PORT 0 RX->TX
0x00000000:0000>lpbk 0 1 S 0
0x00000000:0002>lnkstat
[00] part:00.01 w:cfg[x08]-neg[x00] stk:0.0 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -ACTIVE0 
[04] part:01.01 w:cfg[x08]-neg[x00] stk:0.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[08] part:00.02 w:cfg[x16]-neg[x00] stk:1.0 port_active_lane:xxxxxxxxxxxxxxxx dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[16] part:00.03 w:cfg[x08]-neg[x00] stk:2.0 port_active_lane:0123xxxx         dsp dl_active:0 Rate:G1 LTSSM: Cfg    -US_LW_START 
[20] part:01.02 w:cfg[x08]-neg[x00] stk:2.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[24] part:00.04 w:cfg[x08]-neg[x04] stk:3.0 port_active_lane:0123xxxx         dsp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[28] part:01.03 w:cfg[x08]-neg[x00] stk:3.4 port_active_lane:xxxxxxxx         dsp dl_active:0 Rate:G1 LTSSM: Detect -QUIET 
[32] part:01.00 w:cfg[x16]-neg[x04] stk:4.0 port_active_lane:0123xxxxxxxxxxxx usp dl_active:1 Rate:G4 LTSSM: L0     -L0          
[40] part:00.00 w:cfg[x16]-neg[x08] stk:5.0 port_active_lane:01234567xxxxxxxx usp dl_active:1 Rate:G3 LTSSM: L0     -L0          
[48] part:00.05 w:cfg[x04]-neg[x04] stk:6.0 port_active_lane:0123             dsp dl_active:1 Rate:G4 LTSSM: L0     -L0          

// Enable the generator using PRBS7 on port 16
0x00000000:0000>patgenchk -gen -set 16 0

// Enable the monitor using PRBS7 on port 16
0x00000000:0000>patgenchk -chk -set 16 0

// Read the checker (error) status using PRBS7 on port 16
// Run for each lane (0 to 3 in this link)
0x00000000:0000>patgenchk -chk -get 16 0 
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000000
0x00000000:0000>patgenchk -chk -get 16 1
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000000
0x00000000:0000>patgenchk -chk -get 16 2
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000000
0x00000000:0000>patgenchk -chk -get 16 3
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000000

//INJECT (3) ERRORS ON PORT 0 
0x00000000:000a>patgenchk -err -inj 0 3

// Read the checker (error) status using PRBS7 on port 16
// Run for each lane (0 to 3 in this link)
// Confirm we get three on each lane as injected
0x00000000:0000>patgenchk -chk -get 16 0
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000003
0x00000000:0000>patgenchk -chk -get 16 1
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000003
0x00000000:0000>patgenchk -chk -get 16 2
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000003
0x00000000:0000>patgenchk -chk -get 16 3
Pattern Monitoring: 0
Timestamp: 0s, Error Cnt: 0x00000003
0x00000000:000f>

**Comment by lsgunth** [comment]
I made some progress. It's behaving a bit more like the commands on the uart, but it's still subtly different and I can't get error injection to work on either method.

**Comment by yungwesl** [comment]
Thanks Logan,

Let me put this on my sideburner so I can look into the behavioural differences.  Can you work on some of the other items in The SoW?  I will feedback on this PR once I sort out what might be happening.

**Comment by kelvin-cao** [comment]
Let me know if any test I can do for help.

**Comment by lsgunth** [comment]
Yup, will do.

**Comment by yungwesl** [comment]
Hey Logan,

I think we are 95% there.  The only issue I see from the test case is that the error injection seems to be placing the errors I intended on 3 out of 4 lanes.  This looks like some kind of artifact because I think the MRPC will clear the counter once it is read so it looks like we might be reading lane 0, and then reading lane 0,1,2,3 so lane 0 shows a zero count while lanes 1,2,3 show the proper count

Here I inject 7 errors on Port 0 where Port 16 Tx -> Port 0 (loopback Rx to Tx) -> Port 16 Rx

wesley@kelvin-Gigabyte-right:~/switchtec-user$ sudo ./switchtec diag pattern /dev/i2c-6@0x10 -p 0 -i 7
Injected 7 errors
Port: 0
  Generator: Disabled
  Monitor: PRBS7                0
    Lane 0     Errors: 0x0
    Lane 1     Errors: 0x0
    Lane 2     Errors: 0x0
    Lane 3     Errors: 0x0
    Lane 4     Errors: 0x0
    Lane 5     Errors: 0x0
    Lane 6     Errors: 0x0
    Lane 7     Errors: 0x0
wesley@kelvin-Gigabyte-right:~/switchtec-user$ sudo ./switchtec diag pattern /dev/i2c-6@0x10 -p 16 -m -t PRBS7
Port: 16
  Generator: PRBS7
  Monitor: PRBS7                0
    Lane 0     Errors: 0x0
    Lane 1     Errors: 0x7
    Lane 2     Errors: 0x7
    Lane 3     Errors: 0x7
    Lane 4     Errors: 0x0
    Lane 5     Errors: 0x0
    Lane 6     Errors: 0x0
    Lane 7     Errors: 0x0

**Comment by yungwesl** [comment]
> Hey Logan,
> 
> I think we are 95% there. The only issue I see from the test case is that the error injection seems to be placing the errors I intended on 3 out of 4 lanes. This looks like some kind of artifact because I think the MRPC will clear the counter once it is read so it looks like we might be reading lane 0, and then reading lane 0,1,2,3 so lane 0 shows a zero count while lanes 1,2,3 show the proper count
> 
> Here I inject 7 errors on Port 0 where Port 16 Tx -> Port 0 (loopback Rx to Tx) -> Port 16 Rx
> 
> wesley@kelvin-Gigabyte-right:~/switchtec-user$ sudo ./switchtec diag pattern /dev/i2c-6@0x10 -p 0 -i 7 Injected 7 errors Port: 0 Generator: Disabled Monitor: PRBS7 0 Lane 0 Errors: 0x0 Lane 1 Errors: 0x0 Lane 2 Errors: 0x0 Lane 3 Errors: 0x0 Lane 4 Errors: 0x0 Lane 5 Errors: 0x0 Lane 6 Errors: 0x0 Lane 7 Errors: 0x0 wesley@kelvin-Gigabyte-right:~/switchtec-user$ sudo ./switchtec diag pattern /dev/i2c-6@0x10 -p 16 -m -t PRBS7
> Port: 16
> Generator: PRBS7
> Monitor: PRBS7 0
> Lane 0 Errors: 0x0
> Lane 1 Errors: 0x7
> Lane 2 Errors: 0x7
> Lane 3 Errors: 0x7
> Lane 4 Errors: 0x0
> Lane 5 Errors: 0x0
> Lane 6 Errors: 0x0
> Lane 7 Errors: 0x0

I think we should not print the error counts after an injection but just print the number of injected errors.  The read back needs to happen in a separate atomic command so we do not accidentally clear the counter on a write/then/read sequence

**Comment by lsgunth** [comment]
Ah, I see why that is. The code was querying lane 0 twice: the first time to get the monitor type, the second to get the error count. I've adjusted it so it doesn't have to do that. 

I've also disabled the printing when injecting errors as you asked, but I'm pretty sure this has nothing to do with the issue above. I can probably re-enable the printing if you'd like. It would make the interface more uniform.

None of the reading is atomic, it has to read one lane at a time.

Please give it a try.

**Comment by yungwesl** [comment]
Thanks Logan,

That definitely fixed the 0 count on the readback after injection.

One usage comment:
  **[  --disable, -d ]                    --- Disable all Loopback Modes**

There is a separate command for loopback so is this supposed to be "disable all generators?"

**Comment by lsgunth** [comment]
Yup, just pushed a fix to clean that up.

File: cli/diag.c [hunk 1]
```diff
@@ -110,6 +110,296 @@ static int diag_parse_common_cfg(int argc, char **argv, const char *desc,
...
+		.port_id = -1,
+		.speed = SWITCHTEC_DIAG_LTSSM_GEN4,
+	};
+
+	const struct argconfig_options opts[] = {
+		DEVICE_OPTION,
+		{"port", 'p', "PORT_ID", CFG_NONNEGATIVE, &cfg.port_id,
+		 required_argument, "physical port ID to set/get loopback for"},
+		{"disable", 'd', "", CFG_NONE, &cfg.disable, no_argument,
+		 "Disable all Loopback Modes"},
```
Comment (kelvin-cao) [inline cli/diag.c:187]: Loopback Modes -> loopback modes

File: cli/diag.c [hunk 2]
```diff
@@ -110,6 +110,296 @@ static int diag_parse_common_cfg(int argc, char **argv, const char *desc,
...
+		{"ltssm", 'l', "", CFG_NONE, &cfg.enable_ltssm, no_argument,
+		 "Enable LTSSMloopback mode"},
```
Comment (kelvin-cao) [inline cli/diag.c:189]: LTSSMloopback mode -> LTSSM loopback mode

File: cli/diag.c [hunk 3]
```diff
@@ -110,6 +110,296 @@ static int diag_parse_common_cfg(int argc, char **argv, const char *desc,
...
+	printf("Port: %d\n", port_id);
+	if (gen_pat == SWITCHTEC_DIAG_PATTERN_PRBS_DISABLED)
+		printf("  Generator: Disabled\n");
+	else
+		printf("  Generator: %s\n", pattern_to_str(gen_pat));
+
+	if (mon_pat == SWITCHTEC_DIAG_PATTERN_PRBS_DISABLED) {
+		printf("  Monitor: Disabled\n");
+	} else {
+		printf("  Monitor: %-20s %d\n", pattern_to_str(mon_pat), mon_pat);
```
Comment (kelvin-cao) [inline cli/diag.c:287]: Do we need the 'mon_pat' printed given we already have the string of it printed?

Comment (lsgunth) [reply cli/diag.c]: Oh, that was a debug print... Removed

File: cli/diag.c [hunk 4]
```diff
@@ -110,6 +110,296 @@ static int diag_parse_common_cfg(int argc, char **argv, const char *desc,
...
+		.port_id = -1,
+		.speed = SWITCHTEC_DIAG_LTSSM_GEN4,
+	};
+
+	const struct argconfig_options opts[] = {
+		DEVICE_OPTION,
+		{"port", 'p', "PORT_ID", CFG_NONNEGATIVE, &cfg.port_id,
+		 required_argument, "physical port ID to set/get loopback for"},
+		{"disable", 'd', "", CFG_NONE, &cfg.disable, no_argument,
+		 "Disable all loopback Modes"},
```
Comment (kelvin-cao) [inline cli/diag.c:187]: last nit: Modes -> modes

**Review by kelvin-cao — APPROVED** [review]
Looks good to me except for a minor nit. Thanks.

**Comment by lsgunth** [comment]
Fixed up that minor nit.

**Pull request merged**